helix-term/src/handlers/completion.rs

```diff
@@ -0,0 +1,412 @@
+            .any(|ls| {
+                matches!(&ls.capabilities().completion_provider, Some(lsp::CompletionOptions {
+                        trigger_characters: Some(triggers),
+                        ..
+                    }) if triggers.iter().any(|trigger| rope_ends_with(trigger, text)))
+            });
+
+        let is_auto_trigger = !trigger_char_only
+            && doc
+                .text()
+                .chars_at(primary_cursor)
+                .reversed()
+                .take(config.completion_trigger_len as usize)
+                .all(char_is_word);
+
+        if is_trigger_char || is_auto_trigger {
+            send_blocking(
+                tx,
+                CompletionEvent::Trigger(CompletionTrigger {
+                    trigger_pos: primary_cursor,
+                    doc: doc.id(),
+                    view: view.id,
+                    auto: !is_trigger_char,
+                }),
+            );
+        }
+    }
+}
+
+fn update_completions(cx: &mut commands::Context, c: Option<char>) {
+    cx.callback.push(Box::new(move |compositor, cx| {
+        let editor_view = compositor.find::<ui::EditorView>().unwrap();
+        if let Some(completion) = &mut editor_view.completion {
+            completion.update_filter(c);
+            if completion.is_empty() {
+                editor_view.clear_completion(cx.editor);
+                // clearing completions might mean we want to immediately rerequest them (usually
+                // this occurs if typing a trigger char)
+                if c.is_some() {
+                    trigger_auto_completion(&cx.editor.handlers.completions, cx.editor, false);
+                }
+            }
+        }
+    }))
+}
+
+fn clear_completions(cx: &mut commands::Context) {
+    cx.callback.push(Box::new(|compositor, cx| {
+        let editor_view = compositor.find::<ui::EditorView>().unwrap();
+        editor_view.clear_completion(cx.editor);
+    }))
+}
+
+fn completion_post_command_hook(
+    tx: &Sender<CompletionEvent>,
+    PostCommand { command, cx }: &mut PostCommand<'_, '_>,
+) -> anyhow::Result<()> {
+    if cx.editor.mode == Mode::Insert {
+        if cx.editor.last_completion.is_some() {
+            match command {
+                MappableCommand::Static {
+                    name: "delete_word_forward" | "delete_char_forward" | "completion",
+                    ..
+                } => (),
+                MappableCommand::Static {
+                    name: "delete_char_backward",
+                    ..
+                } => update_completions(cx, None),
+                _ => clear_completions(cx),
+            }
+        } else {
+            let event = match command {
+                MappableCommand::Static {
+                    name:
+                        "delete_word_forward"
+                        | "delete_char_backward"
+                        | "delete_char_forward"
+                        | "completion",
```

Comment:
We're sending a `DeleteText` event when the command is `completion`?

Reply:
I think this is just an artifcat of an earlier implementation, I moved this to a different branch. It doens't matter too much here as long as we don't send a cancel event here (the delete gets ignored since the trigger is gone).